src/handler.ts

```diff
@@ -0,0 +1,685 @@
+
+let counter = 0;
+let parsedVersion: Version;
+
+const checkJava = (
+    java: string,
+): Promise<{
+    run: boolean;
+    message?: ExecException;
+}> =>
+    new Promise((resolve) => {
+        exec(`"${java}" -version`, (error, _, stderr) => {
+            if (error)
+                return resolve({
+                    run: false,
+                    message: error,
+                });
+
+            log(
+                'debug',
+                `Using Java version ${stderr.match(/"(.*?)"/)?.pop()} ${stderr.includes('64-Bit') ? '64-bit' : '32-Bit'}`,
+            );
+            resolve({ run: true });
+        });
+    });
+
+const downloadAsync = async (url: string, directory: string, name: string, retry: boolean, type: string) => {
+    try {
+        mkdirSync(directory, { recursive: true });
+
+        const response = await axios.get(url, {
+            responseType: 'stream',
+            timeout: config.timeout || 50000,
+            httpAgent: new http({ maxSockets: config.maxSockets || 2 }),
+            httpsAgent: new https({ maxSockets: config.maxSockets || 2 }),
+        });
+
+        const totalBytes = parseInt(response.headers['content-length']);
+        let receivedBytes = 0;
+
+        response.data.on('data', (data: Buffer | string) => {
+            typeof data === 'string' ? (receivedBytes += Buffer.byteLength(data)) : (receivedBytes += data.length);
+
+            log('download-status', {
+                name: name,
+                type: type,
+                current: receivedBytes,
+                total: totalBytes,
+            });
+        });
+
+        const file = createWriteStream(join(directory, name));
+        response.data.pipe(file);
+
+        await new Promise((resolve) => {
+            file.on('finish', resolve);
+            file.on('error', async (e) => {
+                log('debug', `Failed to download asset to ${join(directory, name)} due to\n${e}. Retrying... ${retry}`);
+                if (existsSync(join(directory, name))) unlinkSync(join(directory, name));
+                if (retry) await downloadAsync(url, directory, name, false, type);
+                return resolve(e);
+            });
+        });
+
+        log('download', name);
+        return {
+            failed: false,
+            asset: null,
+        };
+    } catch (error) {
+        log('debug', `Failed to download asset to ${join(directory, name)} due\n${error}. Retrying... ${retry}`);
+        if (retry) await downloadAsync(url, directory, name, false, type);
+        return;
+    }
+};
+
+const getVersion = async () => {
+    const versionJsonPath = config.versionJson || join(config.directory!, `${config.version.number}.json`);
+    if (existsSync(versionJsonPath)) {
+        parsedVersion = JSON.parse(readFileSync(versionJsonPath).toString());
+        return parsedVersion;
+    }
+
+    const manifest = `${config.url?.meta}/mc/game/version_manifest.json`;
```

Comment:
Be careful, you might end with an `undefined` at the beginning of the string if not handled properly.